src/engine/PathSearch.cpp

```diff
@@ -0,0 +1,391 @@
+    ids = binding->first->getResult()->idTable().getColumn(binding->second);
+  } else if (isVariable) {
+    return {};
+  } else {
+    ids = std::get<std::vector<Id>>(side);
+  }
+  return ids;
+}
+
+// _____________________________________________________________________________
+std::vector<Path> PathSearch::findPaths(
+    const Id source, const std::unordered_set<uint64_t>& targets,
+    const BinSearchWrapper& binSearch) const {
+  std::forward_list<Edge> edgeStack;
+  Path currentPath;
+  std::unordered_map<uint64_t, std::vector<Path>> pathCache;
+  std::unordered_set<uint64_t> visited;
+
+  auto addToCache =
+      [](std::unordered_map<uint64_t, std::vector<Path>>& pathCache,
+         const Path& path, size_t stopIndex) {
+        for (size_t i = 0; i < stopIndex; i++) {
+          const auto& edge = path.edges_[i];
+          auto startIndex = edge.start_.getBits();
+          pathCache.try_emplace(startIndex, std::vector<Path>());
+          pathCache[startIndex].push_back(path.startingAt(i));
+        }
+      };
+
+  visited.insert(source.getBits());
+  for (auto edge : binSearch.outgoingEdes(source)) {
+    edgeStack.push_front(std::move(edge));
+  }
+
+  while (!edgeStack.empty()) {
+    auto edge = edgeStack.front();
+    edgeStack.pop_front();
+
+    while (!currentPath.empty() && edge.start_ != currentPath.end()) {
+      currentPath.pop_back();
+    }
+
+    currentPath.push_back(edge);
+
+    // TODO clean this up
+    if (pathCache.contains(edge.end_.getBits())) {
+      for (auto subPath : pathCache[edge.end_.getBits()]) {
+        if (subPath.first() == currentPath.first()) {
+          addToCache(pathCache, currentPath, currentPath.size());
+        } else {
+          auto fullPath = currentPath.concat(subPath);
+          addToCache(pathCache, fullPath, currentPath.size());
+        }
+      }
+      continue;
+    }
+
+    if (targets.empty() || targets.contains(edge.end_.getBits())) {
+      addToCache(pathCache, currentPath, currentPath.size());
+    }
+
+    if (!visited.contains(edge.end_.getBits())) {
+      for (auto outgoingEdge : binSearch.outgoingEdes(edge.end_)) {
+        edgeStack.push_front(outgoingEdge);
+      }
+    }
+
+    visited.insert(edge.end_.getBits());
+  }
+
+  return pathCache[source.getBits()];
+}
+
+// _____________________________________________________________________________
+std::vector<Path> PathSearch::allPaths(std::span<const Id> sources,
+                                       std::span<const Id> targets,
+                                       BinSearchWrapper& binSearch) const {
+  std::vector<Path> paths;
+  Path path;
+
+  std::unordered_set<uint64_t> targetSet;
+  for (auto target : targets) {
+    targetSet.insert(target.getBits());
+  }
+
+  if (sources.empty()) {
+    sources = binSearch.getSources();
+  }
+  for (auto source : sources) {
+    for (auto path : findPaths(source, targetSet, binSearch)) {
+      paths.push_back(path);
```

Comment:
This implementation is the one for two separate values clauses, doing the full cartesian product of sources and targets.